### PR TITLE
v3.3.0 release (fixes #4310)

### LINKED ISSUE #4310: release 3.3.0
## Summary

I'd like to start a discussion on this question:

> Should we do a v3.3.0 release of LightGBM some time in June?

## Motivation

My primary motivation for this proposal is that user reports in the last two months have revealed critical issues with the R package (#4007, #4045, #4216, #4259, #4305), and these are fixed by recent PRs related to #3016 (especially #4155 and #4247).

But looking at the PRs that have been merged since LightGBM 3.2.1, there are a lot of other useful fixes, including:

* fix for issue in data parallel learning when feature distributions on partitions don't overlap (#4185)
* more accurate handoffs between C++ side and wrapper packages (#4291, #4292)
* changes to MRO to match scikit-learn's preferences (#3192)
* fix warnings from CUDA builds (#4268)

There has only been one technically "breaking" PR merged, and I think it's ok to include in a minor release (#4197).

## Proposal

I'd like to get other maintainers' thoughts on the following proposal.

1. Do not merge #3234 yet
2. Prepare a 3.3.0 release, to be released some time in June
3. Complete any fixes / feature requests that maintainers suggest before that release

Personally, I'd like to include at least the following in a 3.3.0 release:

* C++
    - [x] #4272
    - [x] incorrect calculation of gamma metric in weighted training (#4174 / #4283)
* Dask
    - [x] eval sets (#4101)
    - [ ] *(not critical)* inconsistencies in leaf counts when performing multiclass classification (#4220)
* Python package
    - [x] fix unexplained jupyter kernel restarts (#4301)
* R package

    - [x] prevent segfaults when objects are serialized / deserialized (#4208)
        - [x] https://github.com/microsoft/LightGBM/pull/4586
    - [x] a deprecation warning for `...` in the R package (https://github.com/microsoft/LightGBM/issues/4226#issuecomment-829473584)
        - [x] https://github.com/microsoft/LightGBM/pull/4571
        - [x] https://github.com/microsoft/LightGBM/pull/4581
        - [x] https://github.com/microsoft/LightGBM/pull/4522
        - [x] https://github.com/microsoft/LightGBM/pull/4523
        - [x] https://github.com/microsoft/LightGBM/pull/4548
        - [x] https://github.com/microsoft/LightGBM/pull/4554
        - [x] https://github.com/microsoft/LightGBM/pull/4572
        - [x] https://github.com/microsoft/LightGBM/pull/4573
    - [x] fix memory leaks (#4282)
        - [x] https://github.com/microsoft/LightGBM/pull/4597
    - [x] prevent session crashes on Windows (#4464)
        - https://github.com/microsoft/LightGBM/pull/4496
    - [x] fix `predict()` being broken when using a file (#4034)
        - https://github.com/microsoft/LightGBM/pull/4545
    - [x] updating how autoconf finds libomp on Mac (#4131)
        - https://github.com/microsoft/LightGBM/pull/4506
        - https://github.com/microsoft/LightGBM/pull/4507

I also think this release could be a good opportunity for maintainers to think carefully about what breaking changes might be made (in addition to #3234) in a 4.0.0 release, and to add deprecation warnings for them in the 3.3.0 release.

<hr>

Thanks for your time and consideration!

### PR DESCRIPTION
**Pull request by jameslamb**
Preparing the 3.3.0 release.

Release checklist below was copied from https://github.com/microsoft/LightGBM/pull/4169.

### Release checklist:

- [x] Update [VERSION.txt](https://github.com/microsoft/LightGBM/blob/master/VERSION.txt) number.
- [x] Update version in [Appveyor config file](https://github.com/microsoft/LightGBM/blob/master/.appveyor.yml).
- [ ] Update version in [`configure` file of R-package](https://github.com/microsoft/LightGBM/blob/master/R-package/configure): `/gha run r-configure`.
- [ ] Run R valgrind checks after all PRs are merged: `/gha run r-valgrind`.
- [ ] Run R Solaris checks after all PRs are merged: `/gha run r-solaris`.
- [ ] Push commit with `v*` tag to trigger `GitHubRelease` action at Azure Pipelines.
- [ ] Convert automatic GitHub release from `Draft` to normal one.
- [ ] Update [`stable` tag at GitHub](https://github.com/microsoft/LightGBM/releases/tag/stable).
- [ ] Trigger [RTD build](https://readthedocs.org/projects/lightgbm/versions/) for new tag.
- [ ] Upload release to [PyPI](https://pypi.org/project/lightgbm/).
- [ ] Upload release to [CRAN](https://cran.r-project.org/package=lightgbm).
- [ ] Update [cran-comments.md](https://github.com/microsoft/LightGBM/blob/master/R-package/cran-comments.md) when new release is accepted on CRAN.
- [ ] Upload release to [NuGet](https://www.nuget.org/packages/LightGBM/).
- [ ] Update version and hash in [Homebrew formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/lightgbm.rb).

### Necessary PRs and other Changes

I don't believe any additional PRs remain (we had been tracking this all along in #4310).

* [ ] Add @shiyu1994 as maintainer in `R-package/DESCRIPTION`

### CODE REVIEW

File: R-package/DESCRIPTION [hunk 1]
```diff
@@ -4,7 +4,8 @@ Title: Light Gradient Boosting Machine
 Version: ~~VERSION~~
 Date: ~~DATE~~
 Authors@R: c(
-    person("Guolin", "Ke", email = "guolin.ke@microsoft.com", role = c("aut", "cre")),
+    person("Yu", "Shi", email = "", role = c("aut", "cre")),
```
Comment (jameslamb) [inline R-package/DESCRIPTION:7]: @shiyu1994 

1. Are you comfortable being the maintainer for the R package, for CRAN purposes? I'm proposing that because I assume Microsoft requires that it be a Microsoft employee.
2. Can you please provide an email address you'd like to use here? This is required.

File: R-package/DESCRIPTION [hunk 2]
```diff
@@ -24,7 +25,8 @@ Authors@R: c(
     person("Alberto", "Ferreira", role = c("ctb")),
     person("Daniel", "Lemire", role = c("ctb")),
     person("Victor", "Zverovich", role = c("cph")),
-    person("IBM Corporation", role = c("ctb"))
+    person("IBM Corporation", role = c("ctb")),
+    person("David", "Cortes", role = c("ctb"))
```
Comment (jameslamb) [inline R-package/DESCRIPTION:29]: proposing adding @david-cortes to DESCRIPTION, thanks to their excellent recent contributions (https://github.com/microsoft/LightGBM/pulls?q=is%3Apr+author%3Adavid-cortes+is%3Aclosed)

File: R-package/DESCRIPTION (hunk 1 above)
Comment (shiyu1994) [reply R-package/DESCRIPTION:7]: Sure. Please use this email address: yushi2@microsoft.com. Thanks!

Comment (jameslamb) [reply R-package/DESCRIPTION:7]:
```suggestion
    person("Yu", "Shi", email = "yushi2@microsoft.com", role = c("aut", "cre")),
```

**Comment by david-cortes** [comment]
I'd like to point out that there's at least one blocker left for a potential CRAN release: https://github.com/microsoft/LightGBM/issues/4598

**Comment by jameslamb** [comment]
> I'd like to point out that there's at least one blocker left for a potential CRAN release: #4598

@david-cortes we are not treating that as a blocker in releasing to CRAN. The code that issue describes has been in every previous LightGBM submission to CRAN and has not been flagged as problematic.

It would be great to address it at some point but I don't favor delaying this release until that issue is picked up.

**Comment by david-cortes** [comment]
> > I'd like to point out that there's at least one blocker left for a potential CRAN release: #4598
> 
> @david-cortes we are not treating that as a blocker in releasing to CRAN. The code that issue describes has been in every previous LightGBM submission to CRAN and has not been flagged as problematic.
> 
> It would be great to address it at some point but I don't favor delaying this release until that issue is picked up.

But that's because it was only accessible from tests that were skipped. As those tests are now unskipped, they'd either have to be set to skipped again, or the issue addressed, otherwise it might randomly fail their tests (particularly solaris).

**Comment by jameslamb** [comment]
> But that's because it was only accessible from tests that were skipped

This is not a correct statement. `LGBM_DatasetCreateFromCSC_R` is invoked dozens of times across many of the R package's tests, since the `agaricus` datasets used for testing are sparse matrices. Any test calling `lgb.Dataset()` or `lgb.train()` or `lgb.cv()` on a dataset built from that data will end up using `LGBM_DatasetCreateFromCSC_R`.

https://github.com/microsoft/LightGBM/blob/a77260f03eb735923125024b0f36df9f63a6e0f3/R-package/tests/testthat/test_basic.R#L10

```r
library(lightgbm)
data(agaricus.train, package = "lightgbm")
class(agaricus.train$data)
```

> [1] "dgCMatrix"
attr(,"package")
[1] "Matrix"

https://github.com/microsoft/LightGBM/blob/a77260f03eb735923125024b0f36df9f63a6e0f3/R-package/R/lgb.Dataset.R#L263-L269

**Review by StrikerRUS — APPROVED** [review]

**Comment by jameslamb** [comment]
@shiyu1994 are you ok with releasing the current state of this branch as 3.3.0? I'd like your approval before we merge this and start pushing to package managers.

**Review by guolinke — APPROVED** [review]

**Comment by shiyu1994** [comment]
Sure. I'm ok with this.

**Review by shiyu1994 — APPROVED** [review]

**Comment by jameslamb** [comment]
Thanks all!

Before we merge... @StrikerRUS did you want https://github.com/microsoft/LightGBM/pull/4637 to make it into 3.3.0? If so we'll need to merge latest `master` here.

**Comment by StrikerRUS** [comment]
Yeah, thanks! Let's merge recent commits.

**Comment by jameslamb** [comment]
Thanks all! I'm running the R Solaris and `valgrind` tests one more time, since changes were merged in that affect the C++ side (#4608, #4643, #4650).

If those tests pass, since we have 4 approvals I'll merge this PR and work through some of the other release items like setting a release / tag here on GitHub and submitting the R package to CRAN.

**Comment by jameslamb** [comment]
It seems that the R Hub jobs are timing out and being killed by GitHub Actions.

Looking at https://yopmail.com?lightgbm_rhub_checks, the most recent messages received there are from September 29, 2021. That suggests to me that the jobs I submitted from comments on this PR are in one of the following states:

* were never successfully submitted to R Hub
* queued at R Hub and will complete at some time in the future
* ran at R Hub but an email notification was not sent

I can write this up in an issue and try to investigate it later, sorry for the delay 😢

**Comment by StrikerRUS** [comment]
I'd like to ask to **not** merge new large PRs with major features for one or two weeks after releasing `v3.3.0`. This will ease the process of creating `v3.3.1` hotfix release if that will be needed in case of some critical bugs in `v3.3.0` we might get reports about.

**Comment by jameslamb** [comment]
Ok, Solaris tests passed! Thanks for your patience, everyone.

I'm going to merge this and start the work on the remaining checklist items.

**Comment by jameslamb** [comment]
GitHub release for 3.3.0 has been published: https://github.com/microsoft/LightGBM/releases/tag/v3.3.0

and tag `v3.3.0` has been pushed: https://github.com/microsoft/LightGBM/tags

here's the Azure Pipelines build triggered by the new tag: https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=11089&view=results

**Comment by jameslamb** [comment]
The jobs at Azure Pipelines completed successfully. Just submitted to CRAN!

image

I did the renaming we talked about in https://github.com/microsoft/LightGBM/pull/4169#issuecomment-818311840 like so:

```shell
wget \
    https://github.com/microsoft/LightGBM/releases/download/v3.3.0/lightgbm-3.3.0-r-cran.tar.gz \
    -O lightgbm_3.3.0.tar.gz
```

@shiyu1994 please check your email. You should get an automated email from CRAN asking you to confirm this submission.

**Comment by jameslamb** [comment]
Great, thank you!

**Comment by jameslamb** [comment]
Source distribution of the R package is now available on CRAN.

https://cran.r-project.org/web/packages/lightgbm/index.html

We'll have to check https://cran.r-project.org/web/checks/check_results_lightgbm.html to see how the automated checks there go. Hoping they succeed!!

**Comment by StrikerRUS** [comment]
@david-cortes Thank you very much!
